README modified (#129)
rsksmart/powpeg-node: README.md

# README

# Welcome to RskJ Powpeg Node

## About

Powpeg node is a specialized rskj node which interacts with both RSK and Bitcoin. This node is used by RSK PowPeg signatories to interact with the Bridge contract and to broadcast peg-out transactions to Bitcoin.

## Software Requirements

1. Java JDK 1.8
2. Bitcoin Core daemon (bitcoind) 0.17v or 0.18v - make sure is the only installation related to bitcoin you have. 
3. Recommended [IntelliJ IDEA](https://www.jetbrains.com/idea/download) as this guide covers the setup with it

> 4. Linux or macOS, this guide covers the setup with the last one

Reviewer: ```suggestion
4. Linux or macOS, this guide covers the setup with the latter option
```

# Welcome to RskJ Powpeg Node

## About

Powpeg node is a specialized rskj node which interacts with both Rootstock and Bitcoin.
This node is used by Rootstock PowPeg signatories to interact with the Bridge contract and to broadcast peg-out transactions to Bitcoin.

## Software Requirements

1. Java JDK 1.8
2. Bitcoin Core daemon (bitcoind) 0.17v or 0.18v
3. A Java compatible IDE. Recommended [IntelliJ IDEA](https://www.jetbrains.com/idea/download) as this guide covers the setup with it

**Not sure how to install any of these? See [software installation help](#software-installation-help)**

## Software installation help
Disclaimer: this documentation will be specific for macOS operative system.

### **Java JDK 1.8**

Although optional we recommend to install jenv to manage different Java versions, to do that run: `brew install jenv`

> Add jenv to your path as it says on the output of jenv brew installation.

Author: we thought it was better not to overcharge the documentation.

> Add jenv to your path as it says on the output of jenv brew installation.

Reviewer 2: It is not a big deal, but I think this is not overcharging the documentation, it is just pointing out the given output that you are mentioning.

But, if you all guys think is better to keep it this way, it's ok for me.

> Download [oracle-jdk8](https://www.oracle.com/ar/java/technologies/javase/javase8-archive-downloads.html) and install it.

Author: we recommended zulu first, but Jose told us it was better to recommend to
download the official java version from Oracle

> Download [oracle-jdk8](https://www.oracle.com/ar/java/technologies/javase/javase8-archive-downloads.html) and install it.

Reviewer 2: Well, the thing is Oracle JDK license is not free and open source(AFAIK), it is limited only to personal use. On the other hand, OpenJDK is free and open source and not limited to personal use. That's my concern. Taking into account that RSK is an open-source project, recommend this instead of the Oracle JDK I think makes more sense.

> Download [oracle-jdk8](https://www.oracle.com/ar/java/technologies/javase/javase8-archive-downloads.html) and install it.

Author: i think you are right. maybe we can leave both

**Note:** *if you have an Apple Silicon chip you may get a notification saying that is not recommended to install a file that is for Intel chips, just ignore it and keep going with the installation.*

Add the just downloaded jdk to the jenv versions, run: `jenv add /Library/Java/JavaVirtualMachines/<ORACLE-JDK8-DIRECTORY>/Contents/Home/`

If you are using another java version in other projects, run `jenv local 1.8.0.352`. Otherwise you can set it globally by running: `jenv global 1.8.0.352`

To check that it worked, open a new terminal instance and run `java -version` command. It should output something like 

```
openjdk version "1.8.0_352"
OpenJDK Runtime Environment
OpenJDK 64-Bit Server VM 

> ```

Author: you are right, the output went obsolete. i will change it

*If not, try closing the terminal and open a new one cause sometimes it stays cached.*

### Bitcoind

To install version `0.18.0`

```bash
curl -O https://bitcoin.org/bin/bitcoin-core-0.18.0/bitcoin-0.18.0-osx64.tar.gz
tar -zxf bitcoin-0.18.0-osx64.tar.gz
sudo mkdir -p /usr/local/bin
sudo cp bitcoin-0.18.0/bin/bitcoin* /usr/local/bin/.

> rm -rf bitcoin-0.18.0

Reviewer 2: This command is not necessary. It is just deleting the folder you just unzip.

Author: ok! ill delete the rm -rf line

```

To validate it run `bitcoind -daemon`.

It should output ```Bitcoin server starting```

Run `bitcoin-cli stop` afterwards.

## Setting up the project
Create a directory (for example, “powpeg-project”) to hold the rskj node, the powpeg node and further configurations.

### Fetch the code

Inside the `powpeg-project` directory clone the powpeg-node repository.

```bash
git clone https://github.com/rsksmart/powpeg-node
```

```bash
cd powpeg-node
```

### Verify the code

Before anything, you must ensure the security chain of the source code. For that, you must go through the following steps. For Linux based OS (Ubuntu for example) it’s recommended install `gnupg-curl` to download the key through HTTPS.

1. Download sec channel public key

```bash

> gpg --keyserver https://secchannel.rsk.co/SUPPORT.asc --recv-keys A6DBEAC640C5A14B

Reviewer 2: This command does not work for me. Is working for you?

```bash
> gpg --keyserver https://secchannel.rsk.co/SUPPORT.asc --recv-keys A6DBEAC640C5A14B
gpg: keyserver receive failed: Network is unreachable
```

Author: you are right, it doesn't work for us either right now. we will try to understand why and solve it

# Welcome to RskJ Powpeg Node

## About

Powpeg node is a specialized rskj node which interacts with both Rootstock and Bitcoin.
This node is used by Rootstock PowPeg signatories to interact with the Bridge contract and to broadcast peg-out transactions to Bitcoin.

## Software Requirements

1. Java JDK 1.8
2. Bitcoin Core daemon (bitcoind) 0.17v or 0.18v
3. A Java compatible IDE. Recommended [IntelliJ IDEA](https://www.jetbrains.com/idea/download) as this guide covers the setup with it

**Not sure how to install any of these? See [software installation help](#software-installation-help)**

## Software installation help

> Disclaimer: this documentation will be specific for macOS operative system.

Reviewer: ```suggestion
Disclaimer: this documentation will be specific for macOS operating system.
```

### **Java JDK 1.8**

Although optional we recommend to install jenv to manage different Java versions, to do that run: `brew install jenv`

Add jenv to your path as it says on the output of jenv brew installation.

```
To activate jenv, add the following to your ~/.zshrc:
  export PATH="$HOME/.jenv/bin:$PATH"
  eval "$(jenv init -)"
```

Download [oracle-jdk8](https://www.oracle.com/ar/java/technologies/javase/javase8-archive-downloads.html) and install it.

> **Note:** *if you have an Apple Silicon chip you may get a notification saying that is not recommended to install a file that is for Intel chips, just ignore it and keep going with the installation.*

Reviewer: ```suggestion
**Note:** *if you have an Apple Silicon chip you may get a notification saying that it is not recommended to install a file that is for Intel chips, just ignore it and keep going with the installation.*
```

To check that it worked, open a new terminal instance and run `java -version` command. It should output something like 

```
java version "1.8.0_351"
Java(TM) SE Runtime Environment (build 1.8.0_351-b10)
Java HotSpot(TM) 64-Bit Server VM (build 25.351-b10, mixed mode)
```

*If not, try closing the terminal and open a new one cause sometimes it stays cached.*

### Bitcoind

To install version `0.18.0`

```bash
curl -O https://bitcoin.org/bin/bitcoin-core-0.18.0/bitcoin-0.18.0-osx64.tar.gz
tar -zxf bitcoin-0.18.0-osx64.tar.gz
sudo mkdir -p /usr/local/bin
sudo cp bitcoin-0.18.0/bin/bitcoin* /usr/local/bin/.
```

To validate it run `bitcoind -daemon`.

It should output ```Bitcoin server starting```

Run `bitcoin-cli stop` afterwards.

## Setting up the project
Create a directory (for example, “powpeg-project”) to hold the rskj node, the powpeg node and further configurations.

### Fetch the code

Inside the `powpeg-project` directory clone the powpeg-node repository.

```bash
git clone https://github.com/rsksmart/powpeg-node
```

```bash
cd powpeg-node
```

### Verify the code

> Before anything, you must ensure the security chain of the source code. For that, you must go through the following steps. For Linux based OS (Ubuntu for example) it’s recommended install `gnupg-curl` to download the key through HTTPS.

Reviewer: ```suggestion
Before anything, you must ensure the security chain of the source code. For that, you must go through the following steps. For Linux based OS (Ubuntu for example) it’s recommended to install `gnupg-curl` to download the key through HTTPS.
```

You should get something like:

```
configure.sh: OK
sha256sum: WARNING: 19 lines are improperly formatted
```

**MacOS:**

```bash
shasum --check SHA256SUMS.asc
```

You should get something like:

```
configure.sh: OK
sha256sum: WARNING: 19 lines are improperly formatted
```

### Compile the code!

Run configure script to configure secure environment.

```bash
./configure.sh
```

### Import and configure the project

To import the project to IntelliJ IDEA go to `File > New > Project from existing sources...` Select `powpeg-node/build.gradle` and import.

- **Create a new running configuration with name FedRunner** (go to `Run > Edit configurations > Add configuration > Application`)
- **Module*:*** `-cp federate-node.main`

*When you select the module, the java version should change automatically*

- **Main Class:** `co.rsk.federate.FederateRunner`
- **Program Arguments:** add `--regtest -–reset`

*Remove `--reset` if you want to keep your database after restarting the node*

- **VM options:** `-Drsk.conf.file=/<PATH_TO_CONF_FILE>` (to add VM options, go to `Modify Options > Add VM options`)

### Build

Then clean and build project using:

```bash
./gradlew clean build
```

> ### Scripts

Reviewer 3: ```suggestion
### Running bitcoind
```

Create a new directory called datadir inside powpeg-project. In the next step you need to replace `<PATH_TO_DATA_DIR>` with this directory absolute path.

Create a new directory called scripts inside powpeg-project with the following scripts inside:

File `bitcoin-node.sh`

```bash
#!/bin/bash
bitcoind -regtest -printtoconsole -server -rpcuser=rsk -rpcpassword=rsk -rpcport=18332 -txindex -debug=net -deprecatedrpc=signrawtransaction -addresstype=legacy -deprecatedrpc=accounts -deprecatedrpc=generate -datadir=<PATH_TO_DATA_DIR>
```

File `btc-regtest.sh`

```bash
#!/bin/bash
bitcoin-cli -regtest --rpcuser=rsk --rpcpassword=rsk --rpcport=18332 $@
```

Start the Bitcoin node with (use another terminal as you will have to keep it running)

```bash
./scripts/bitcoin-node.sh
```

Generate 1 block with

```bash
 ./scripts/btc-regtest.sh generate 1
```

This should return the block hash as an object.

Get the amount of blocks available with

```bash
./scripts/btc-regtest getblockcount
```

> This should return an integer.

Reviewer 3: We could also add this link as reference https://github.com/BlockchainCommons/Learning-Bitcoin-from-the-Command-Line
